pageserver/src/tenant/timeline/compaction.rs

```diff
@@ -501,8 +489,11 @@ impl Timeline {
 
         for &DeltaEntry { key: next_key, .. } in all_keys.iter() {
             if let Some(prev_key) = prev {
-                // just first fast filter
-                if next_key.to_i128() - prev_key.to_i128() >= min_hole_range {
+                // just first fast filter, do not create hole entries for metadata keys. The last hole in the
+                // compaction is the gap between data key and metadata keys.
+                if next_key.to_i128() - prev_key.to_i128() >= min_hole_range
+                    && !Key::is_metadata_key(&prev_key)
+                {
```

Comment:
Not sure I understand what's going on here. Do you have an example?

Reply:
For metadata keys the keyspace is always sparse, and the algorithm will try to create too many holes / partitions. We just skip doing partition for metadata keys here so that it's always handled as a whole in the compaction and image layer creations.